src/Raven.Server/Documents/DatabasesLandlord.cs

```diff
@@ -341,12 +397,78 @@ public bool ShouldDeleteDatabase(TransactionOperationContext context, string dbN
 
             // We materialize the values here because we close the read transaction
             var record = rawRecord.MaterializedRecord;
-            context.CloseTransaction();
+
+            if (sharded == false)
+            {
+                context.CloseTransaction();
+            }
 
             DeleteDatabase(dbName, deletionInProgress, record);
+
+            if (sharded)
+            {
+                // we need to remove the sharded context from resource cache, after we remove all the shards
+                if (OnShardedDatabaseDelete == null)
+                    OnShardedDatabaseDelete += DatabasesLandlord_OnShardedDatabaseDelete;
+
+                // need to remove the database from _shardedDatabases resource cache
+                RemoveShardedDatabase(dbName);
+            }
+            
             return true;
         }
 
+        private void DatabasesLandlord_OnShardedDatabaseDelete(string name, TransactionOperationContext context)
```

Comment:
I changed this code, I have to keep the context because other shards use the same context as well, the context will be disposed after we delete the sharded context